php/src/Communicator.cpp

```diff
@@ -223,15 +217,42 @@ RegisteredCommunicatorMap _registeredCommunicators;
 
 // std::mutex constructor is constexpr so it is statically initialized
 std::mutex _registeredCommunicatorsMutex;
-
-IceUtil::TimerPtr _timer;
+std::condition_variable _reapCond;
+std::thread _reapThread;
+bool _reapingFinished = true;
 
 //
 // This map is stored in the "global" variables for each PHP request and holds
 // the communicators that have been created (or registered communicators that have
 // been used) by the request.
 //
 typedef map<Ice::CommunicatorPtr, CommunicatorInfoIPtr> CommunicatorMap;
+
+void
+reapRegisteredCommunicators()
+{
+    // This function must be called with _registeredCommunicatorsMutex locked
+    auto now = std::chrono::steady_clock::now();
+    RegisteredCommunicatorMap::iterator p = _registeredCommunicators.begin();
+    while(p != _registeredCommunicators.end())
+    {
+        if(p->second->lastAccess + std::chrono::hours(p->second->expires) <= now)
+        {
+            try
+            {
+                p->second->communicator->destroy();
+            }
+            catch(...)
+            {
+            }
+            _registeredCommunicators.erase(p++);
```

Comment:
Is this the same as

```
p = _registeredCommunicators.erase(p);
```

Reply:
This was there from before my change. I can update it though

Reply:
Actually, I think it's clearer written this way.